xarray/plot/dataset_plot.py

```diff
@@ -422,7 +422,7 @@ def scatter(ds, x, y, ax, **kwargs):
 
     if hue_style == "discrete":
         primitive = []
-        for label in np.unique(data["hue"].values):
+        for label in pd.unique(data["hue"].values.ravel()):
```

Comment:
is the `ravel` the bit that matters? I don't see why switching from numpy to pandas fixes it. `pd.unique` claims it's significantly faster than numpy so the change could be good in any case.

Reply:
the bit that matters is `pd.unique`, which does not sort the result. `ravel` / `flatten` is required because `pd.unique` only accepts 1D data.

Reply:
IIUC the order should not matter because we pass the `label` kwarg to scatter which should associate the right labels with the right markers.

The legend is created here: https://github.com/pydata/xarray/blob/bf0fe2caca1d2ebc4f1298f019758baa12f68b94/xarray/plot/dataset_plot.py#L339-L340 and follows the first example here: https://matplotlib.org/3.1.1/gallery/lines_bars_and_markers/scatter_with_legend.html

Reply:
that works for normal plots, although `np.unique` changes the order (see the second example in https://github.com/pydata/xarray/issues/4641#issuecomment-737853069). For facetgrids, however, the labels in the legend are pulled out directly from the data and thus are not sorted (look at the values passed as `labels`): https://github.com/pydata/xarray/blob/bf0fe2caca1d2ebc4f1298f019758baa12f68b94/xarray/plot/facetgrid.py#L382-L388
so we have the choice between sorting both and not sorting at all.

Reply:
In that case, I think the right solution is to pass the `label` kwarg in the `map_*` functions of facetgrid and not set the `labels` kwarg when making the legend. That way it's never out of sync.

Reply:
that seems to be a bigger change and might be related to this comment: https://github.com/pydata/xarray/blob/ed255736664f8f0b4ea199c8f91bffaa89522d03/xarray/plot/facetgrid.py#L352-L353

Reply:
True. OK let's add some tests and merge?

Reply:
done, I think